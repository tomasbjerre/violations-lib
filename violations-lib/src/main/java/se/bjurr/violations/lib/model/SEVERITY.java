package se.bjurr.violations.lib.model;

public enum SEVERITY {
  INFO,
  WARN,
  ERROR
}
