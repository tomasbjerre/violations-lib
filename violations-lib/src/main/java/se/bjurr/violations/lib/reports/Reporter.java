package se.bjurr.violations.lib.reports;

public enum Reporter {
  ARMGCC("ARM-GCC", Parser.CLANG, "https://developer.arm.com/open-source/gnu-toolchain/gnu-rm", ""),
  ANDROIDLINT(
      "AndroidLint", Parser.ANDROIDLINT, "http://developer.android.com/tools/help/lint.html", ""),
  ANSIBLELATER(
      "Ansible-Later",
      Parser.ANSIBLELATER,
      "https://github.com/thegeeklab/ansible-later",
      "With `json` format"),
  ANSIBLELINT(
      "AnsibleLint", Parser.FLAKE8, "https://github.com/willthames/ansible-lint", "With `-p`"),
  BANDIT(
      "Bandit",
      Parser.CLANG,
      "https://github.com/PyCQA/bandit",
      "With `bandit -r examples/ -f custom -o bandit.out --msg-template \"{abspath}:{line}: {severity}: {test_id}: {msg}\"`"),
  CLANG("CLang", Parser.CLANG, "https://clang-analyzer.llvm.org/", ""),
  COVERITY("Coverity", Parser.COVERITY, "https://scan.coverity.com/", ""),
  CFN(
      "CloudFormation Linter",
      Parser.JUNIT,
      "https://github.com/aws-cloudformation/cfn-lint",
      "`cfn-lint . -f junit --output-file report-junit.xml`"),
  CPD("CPD", Parser.CPD, "http://pmd.sourceforge.net/pmd-4.3.0/cpd.html", ""),
  CPPCHECK(
      "CPPCheck",
      Parser.CPPCHECK,
      "http://cppcheck.sourceforge.net/",
      "With `cppcheck test.cpp --output-file=cppcheck.xml --xml`"),
  CPPLINT("CPPLint", Parser.CPPLINT, "https://github.com/theandrewdavis/cpplint", ""),
  CSSLINT("CSSLint", Parser.CSSLINT, "https://github.com/CSSLint/csslint", ""),
  CHECKSTYLE("Checkstyle", Parser.CHECKSTYLE, "http://checkstyle.sourceforge.net/", ""),
  CODENARC("CodeNarc", Parser.CODENARC, "http://codenarc.sourceforge.net/", ""),
  CODECLIMATE("CodeClimate", Parser.CODECLIMATE, "https://codeclimate.com/", ""),
  DEPENDENCYCHECK(
      "Dependency Check",
      Parser.SARIF,
      "https://jeremylong.github.io/DependencyCheck/",
      "Using `--format SARIF`"),
  DETEKT(
      "Detekt",
      Parser.CHECKSTYLE,
      "https://github.com/arturbosch/detekt",
      "With `--output-format xml`."),
  ERB(
      "ERB",
      Parser.CLANG,
      "https://www.puppetcookbook.com/posts/erb-template-validation.html",
      "With `erb -P -x -T '-' \"${it}\" | ruby -c 2>&1 >/dev/null | grep '^-' | sed -E 's/^-([a-zA-Z0-9:]+)/${filename}\\1 ERROR:/p' > erbfiles.out`."),
  DOCFX("DocFX", Parser.DOCFX, "http://dotnet.github.io/docfx/", ""),
  DOXYGEN("Doxygen", Parser.CLANG, "https://www.stack.nl/~dimitri/doxygen/", ""),
  ESLINT(
      "ESLint",
      Parser.CHECKSTYLE,
      "https://github.com/sindresorhus/grunt-eslint",
      "With `format: 'checkstyle'`."),
  FINDBUGS("Findbugs", Parser.FINDBUGS, "http://findbugs.sourceforge.net/", ""),
  FLAKE8("Flake8", Parser.FLAKE8, "http://flake8.readthedocs.org/en/latest/", ""),
  DART("Dart", Parser.MACHINE, "https://dart.dev/", "With `dart analyze --format=machine`"),
  FXCOP("FxCop", Parser.FXCOP, "https://en.wikipedia.org/wiki/FxCop", ""),
  GCC("GCC", Parser.CLANG, "https://gcc.gnu.org/", ""),
  GHS("GHS", Parser.GHS, "https://www.ghs.com/", ""),
  GENDARME(
      "Gendarme",
      Parser.GENDARME,
      "http://www.mono-project.com/docs/tools+libraries/tools/gendarme/",
      ""),
  GENERIC(
      "Generic reporter",
      Parser.GENERIC,
      "",
      "Will create one single violation with all the content as message."),
  GOLINT("GoLint", Parser.GOLINT, "https://github.com/golang/lint", ""),
  GOLANGLINT(
      "GolangCI-Lint",
      Parser.CHECKSTYLE,
      "https://github.com/golangci/golangci-lint/",
      "With `--out-format=checkstyle`."),
  GOVET("GoVet", Parser.GOLINT, "https://golang.org/cmd/vet/", "Same format as GoLint."),
  GOOGLEERRORPRONE(
      "GoogleErrorProne", Parser.GOOGLEERRORPRONE, "https://github.com/google/error-prone", ""),
  HADOLINT(
      "HadoLint",
      Parser.CHECKSTYLE,
      "https://github.com/hadolint/hadolint/",
      "With `-f checkstyle`"),
  INFER("Infer", Parser.PMD, "http://fbinfer.com/", "Facebook Infer. With `--pmd-xml`."),
  IAR(
      "IAR",
      Parser.IAR,
      "https://www.iar.com/iar-embedded-workbench/",
      "With `--no_wrap_diagnostics`"),
  JACOCO("JACOCO", Parser.JACOCO, "https://www.jacoco.org/", ""),
  JCREPORT("JCReport", Parser.JCREPORT, "https://github.com/jCoderZ/fawkez/wiki/JcReport", ""),
  JSHINT(
      "JSHint",
      Parser.JSLINT,
      "http://jshint.com/",
      "With `--reporter=jslint` or the CHECKSTYLE parser with `--reporter=checkstyle`"),
  JUNIT("JUnit", Parser.JUNIT, "https://junit.org/junit4/", "It only contains the failures."),
  KTLINT("KTLint", Parser.CHECKSTYLE, "https://github.com/shyiko/ktlint", ""),
  KLOCWORK(
      "Klocwork",
      Parser.KLOCWORK,
      "http://www.klocwork.com/products-services/klocwork/static-code-analysis",
      ""),
  KOTLINGRADLE(
      "KotlinGradle",
      Parser.KOTLINGRADLE,
      "https://github.com/JetBrains/kotlin",
      "Output from Kotlin Gradle Plugin."),
  KOTLINMAVEN(
      "KotlinMaven",
      Parser.KOTLINMAVEN,
      "https://github.com/JetBrains/kotlin",
      "Output from Kotlin Maven Plugin."),
  LINT("Lint", Parser.LINT, "", "A common XML format, used by different linters."),
  MCCABE("Mccabe", Parser.FLAKE8, "https://pypi.python.org/pypi/mccabe", ""),
  MYPY("MyPy", Parser.MYPY, "https://pypi.python.org/pypi/mypy-lang", ""),
  MSBUILDLOG(
      "MSBuildLog",
      Parser.MSBULDLOG,
      "https://docs.microsoft.com/en-us/visualstudio/msbuild/obtaining-build-logs-with-msbuild?view=vs-2019",
      "With `-fileLogger` use `.*msbuild\\\\.log$` as pattern or `-fl -flp:logfile=MyProjectOutput.log;verbosity=diagnostic` for a custom output filename"),
  MSCPP("MSCpp", Parser.MSCPP, "https://visualstudio.microsoft.com/vs/features/cplusplus/", ""),
  NULLAWAY(
      "NullAway",
      Parser.GOOGLEERRORPRONE,
      "https://github.com/uber/NullAway",
      "Same format as Google Error Prone."),
  PCLINT(
      "PCLint",
      Parser.PCLINT,
      "http://www.gimpel.com/html/pcl.htm",
      "PC-Lint using the same output format as the Jenkins warnings plugin, [_details here_](https://wiki.jenkins.io/display/JENKINS/PcLint+options)"),
  PHPCS(
      "PHPCS",
      Parser.CHECKSTYLE,
      "https://github.com/squizlabs/PHP_CodeSniffer",
      "With `phpcs api.php --report=checkstyle`."),
  PHPPMD("PHPPMD", Parser.PMD, "https://phpmd.org/", "With `phpmd api.php xml ruleset.xml`."),
  PMD("PMD", Parser.PMD, "https://pmd.github.io/", ""),
  PEP8("Pep8", Parser.FLAKE8, "https://github.com/PyCQA/pycodestyle", ""),
  PERLCRITIC("PerlCritic", Parser.PERLCRITIC, "https://github.com/Perl-Critic", ""),
  PITEST("PiTest", Parser.PITEST, "http://pitest.org/", ""),
  PROTOLINT("ProtoLint", Parser.PROTOLINT, "https://github.com/yoheimuta/protolint", ""),
  PUPPETLINT(
      "Puppet-Lint",
      Parser.CLANG,
      "http://puppet-lint.com/",
      "With `-log-format %{fullpath}:%{line}:%{column}: %{kind}: %{message}`"),
  PYDOCSTYLE("PyDocStyle", Parser.PYDOCSTYLE, "https://pypi.python.org/pypi/pydocstyle", ""),
  PYFLAKES("PyFlakes", Parser.FLAKE8, "https://pypi.python.org/pypi/pyflakes", ""),
  PYLINT(
      "PyLint",
      Parser.PYLINT,
      "https://www.pylint.org/",
      "With `pylint --output-format=parseable`."),
  RESHARPER("ReSharper", Parser.RESHARPER, "https://www.jetbrains.com/resharper/", ""),
  RUBYCOP(
      "RubyCop",
      Parser.CLANG,
      "http://rubocop.readthedocs.io/en/latest/formatters/",
      "With `rubycop -f clang file.rb`"),
  SARIF(
      "SARIF",
      Parser.SARIF,
      "https://github.com/oasis-tcs/sarif-spec",
      "v2.x. Microsoft Visual C# can generate it with `ErrorLog=\"BuildErrors.sarif,version=2\"`."),
  SBTSCALAC("SbtScalac", Parser.SBTSCALAC, "http://www.scala-sbt.org/", ""),
  SEMGREP("Semgrep", Parser.SEMGREP, "https://semgrep.dev/", "With `--json`."),
  SIMIAN("Simian", Parser.SIMIAN, "http://www.harukizaemon.com/simian/", ""),
  SPOTBUGS("Spotbugs", Parser.FINDBUGS, "https://spotbugs.github.io/", ""),
  STYLECOP("StyleCop", Parser.STYLECOP, "https://stylecop.codeplex.com/", ""),
  SWIFTLINT(
      "SwiftLint",
      Parser.CHECKSTYLE,
      "https://github.com/realm/SwiftLint",
      "With `--reporter checkstyle`."),
  SONAR(
      "Sonar",
      Parser.SONAR,
      "https://www.sonarqube.org/",
      "With `mvn sonar:sonar -Dsonar.analysis.mode=preview -Dsonar.report.export.path=sonar-report.json`."
          + " Removed in 7.7, see [SONAR-11670](https://jira.sonarsource.com/browse/SONAR-11670) but can be retrieved with:"
          + " `curl --silent 'http://sonar-server/api/issues/search?componentKeys=unique-key&resolved=false' | jq -f sonar-report-builder.jq > sonar-report.json`."),
  TSLINT(
      "TSLint",
      Parser.CHECKSTYLE,
      "https://palantir.github.io/tslint/usage/cli/",
      "With `-t checkstyle`"),
  VALGRIND("Valgrind", Parser.VALGRIND, "https://valgrind.org/", "With `--xml=yes`."),
  XUNIT("XUnit", Parser.XUNIT, "https://xunit.net/", "It only contains the failures."),
  XMLLINT("XMLLint", Parser.XMLLINT, "http://xmlsoft.org/xmllint.html", ""),
  YAMLLINT(
      "YAMLLint",
      Parser.YAMLLINT,
      "https://yamllint.readthedocs.io/en/stable/index.html",
      "With `-f parsable`"),
  ZPTLINT("ZPTLint", Parser.ZPTLINT, "https://pypi.python.org/pypi/zptlint", ""),
  SCALASTYLE("Scalastyle", Parser.CHECKSTYLE, "http://www.scalastyle.org/", "");

  private final String url;
  private final String name;
  private final Parser parser;
  private final String note;

  Reporter(final String name, final Parser parser, final String url, final String note) {
    this.name = name;
    this.parser = parser;
    this.url = url;
    this.note = note;
  }

  public Parser getParser() {
    return this.parser;
  }

  public String getName() {
    return this.name;
  }

  public String getUrl() {
    return this.url;
  }

  public String getNote() {
    return this.note;
  }
}
