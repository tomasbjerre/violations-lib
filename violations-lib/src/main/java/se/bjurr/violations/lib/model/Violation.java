package se.bjurr.violations.lib.model;

import static se.bjurr.violations.lib.util.Utils.checkNotNull;
import static se.bjurr.violations.lib.util.Utils.emptyToNull;
import static se.bjurr.violations.lib.util.Utils.firstNonNull;
import static se.bjurr.violations.lib.util.Utils.nullToEmpty;

import java.io.Serializable;
import java.util.HashMap;
import java.util.Map;
import se.bjurr.violations.lib.reports.Parser;

public class Violation implements Serializable, Comparable<Violation> {

  private static final long serialVersionUID = -6052921679385466168L;

  /** A {@link #file} used when there is no file specified in the parsed report. */
  public static final String NO_FILE = "-";

  /** A {@link #startLine} used when there is no line specified in the parsed report. */
  public static final Integer NO_LINE = 0;

  private Integer column;
  private final Integer endLine;
  private final Integer endColumn;

  /** Also see {@link #NO_FILE} */
  private final String file;

  private final String message;

  /** The algorithm, the format, used. */
  private final Parser parser;

  /**
   * Intended as the tool used to find the violation. Like Detekt, when it is being used to find
   * violations and report them in the {@link Parser#CHECKSTYLE} format.
   */
  private String reporter;

  private final String rule;
  private final String category;

  /**
   * Something that identifies a group that this violation belongs to. First introduced with {@link
   * se.bjurr.violations.lib.parsers.CPPCheckParser} to record what error tag each violation belongs
   * to.
   */
  private final String group;

  private final SEVERITY severity;
  private final String source;
  private final Map<String, String> specifics;
  private final Integer startLine;

  public static class ViolationBuilder {

    private Integer column;
    private Integer endLine;
    private Integer endColumn;
    private String file;
    private String message;
    private Parser parser;
    private String reporter;
    private String rule;
    private String category;
    private SEVERITY severity;
    private String source;
    private Map<String, String> specifics = new HashMap<>();
    private Integer startLine;
    private String group;

    private ViolationBuilder() {}

    public Violation build() {
      return new Violation(this);
    }

    public ViolationBuilder setColumn(final Integer column) {
      this.column = column;
      return this;
    }

    public ViolationBuilder setEndLine(final Integer endLine) {
      this.endLine = endLine;
      return this;
    }

    public ViolationBuilder setEndColumn(final Integer endColumn) {
      this.endColumn = endColumn;
      return this;
    }

    public ViolationBuilder setFile(final String file) {
      this.file = file;
      return this;
    }

    public ViolationBuilder setMessage(final String message) {
      this.message = message;
      return this;
    }

    public ViolationBuilder setParser(final Parser parser) {
      this.parser = parser;
      return this;
    }

    public ViolationBuilder setReporter(final String reporter) {
      this.reporter = reporter;
      return this;
    }

    public ViolationBuilder setRule(final String rule) {
      this.rule = rule;
      return this;
    }

    public ViolationBuilder setCategory(final String category) {
      this.category = category;
      return this;
    }

    public ViolationBuilder setSeverity(final SEVERITY severity) {
      this.severity = severity;
      return this;
    }

    public ViolationBuilder setSource(final String source) {
      this.source = source;
      return this;
    }

    public ViolationBuilder setSpecific(final String specificsKey, final Integer specificsValue) {
      this.specifics.put(specificsKey, Integer.toString(specificsValue));
      return this;
    }

    public ViolationBuilder setSpecific(final String specificsKey, final String specificsValue) {
      this.specifics.put(specificsKey, specificsValue);
      return this;
    }

    public ViolationBuilder setSpecifics(final Map<String, String> specifics) {
      this.specifics = specifics;
      return this;
    }

    public ViolationBuilder setStartLine(final Integer startLine) {
      this.startLine = startLine;
      return this;
    }

    public ViolationBuilder setGroup(final String group) {
      this.group = group;
      return this;
    }
  }

  public static ViolationBuilder violationBuilder() {
    return new ViolationBuilder();
  }

  public Violation() {
    this.startLine = null;
    this.endLine = null;
    this.endColumn = null;
    this.severity = null;
    this.message = null;
    this.file = null;
    this.source = null;
    this.rule = null;
    this.category = null;
    this.reporter = null;
    this.specifics = null;
    this.parser = null;
    this.group = null;
  }

  public Violation(final ViolationBuilder vb) {
    this.parser = checkNotNull(vb.parser, "reporter");
    if (vb.reporter != null && !vb.reporter.trim().isEmpty()) {
      this.reporter = vb.reporter;
    } else {
      this.reporter = this.parser.name();
    }
    this.startLine = checkNotNull(vb.startLine, "startline");
    this.endLine = firstNonNull(vb.endLine, vb.startLine);
    this.column = vb.column;
    this.endColumn = vb.endColumn;
    this.severity = checkNotNull(vb.severity, "severity");
    this.message = checkNotNull(emptyToNull(vb.message), "message");
    this.file = frontSlashes(checkNotNull(emptyToNull(vb.file), "file"));
    this.source = nullToEmpty(vb.source);
    this.rule = nullToEmpty(vb.rule);
    this.category = nullToEmpty(vb.category);
    this.group = nullToEmpty(vb.group);
    this.specifics = vb.specifics;
  }

  public static String frontSlashes(final String in) {
    return in.replace("\\\\", "\\").replaceAll("\\\\", "/");
  }

  public Violation(final Violation v) {
    this.parser = v.parser;
    this.reporter = v.reporter;
    this.startLine = v.startLine;
    this.endLine = v.endLine;
    this.endColumn = v.endColumn;
    this.column = v.column;
    this.severity = v.severity;
    this.message = v.message;
    this.file = v.file;
    this.source = v.source;
    this.rule = v.rule;
    this.category = v.category;
    this.group = v.group;
    this.specifics = v.specifics;
  }

  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if (obj == null) {
      return false;
    }
    if (this.getClass() != obj.getClass()) {
      return false;
    }
    final Violation other = (Violation) obj;
    if (this.category == null) {
      if (other.category != null) {
        return false;
      }
    } else if (!this.category.equals(other.category)) {
      return false;
    }
    if (this.column == null) {
      if (other.column != null) {
        return false;
      }
    } else if (!this.column.equals(other.column)) {
      return false;
    }
    if (this.endLine == null) {
      if (other.endLine != null) {
        return false;
      }
    } else if (!this.endLine.equals(other.endLine)) {
      return false;
    }
    if (this.endColumn == null) {
      if (other.endColumn != null) {
        return false;
      }
    } else if (!this.endColumn.equals(other.endColumn)) {
      return false;
    }
    if (this.file == null) {
      if (other.file != null) {
        return false;
      }
    } else if (!this.file.equals(other.file)) {
      return false;
    }
    if (this.group == null) {
      if (other.group != null) {
        return false;
      }
    } else if (!this.group.equals(other.group)) {
      return false;
    }
    if (this.message == null) {
      if (other.message != null) {
        return false;
      }
    } else if (!this.message.equals(other.message)) {
      return false;
    }
    if (this.parser != other.parser) {
      return false;
    }
    if (this.reporter == null) {
      if (other.reporter != null) {
        return false;
      }
    } else if (!this.reporter.equals(other.reporter)) {
      return false;
    }
    if (this.rule == null) {
      if (other.rule != null) {
        return false;
      }
    } else if (!this.rule.equals(other.rule)) {
      return false;
    }
    if (this.severity != other.severity) {
      return false;
    }
    if (this.source == null) {
      if (other.source != null) {
        return false;
      }
    } else if (!this.source.equals(other.source)) {
      return false;
    }
    if (this.specifics == null) {
      if (other.specifics != null) {
        return false;
      }
    } else if (!this.specifics.equals(other.specifics)) {
      return false;
    }
    if (this.startLine == null) {
      if (other.startLine != null) {
        return false;
      }
    } else if (!this.startLine.equals(other.startLine)) {
      return false;
    }
    return true;
  }

  public Integer getColumn() {
    return firstNonNull(this.column, -1);
  }

  public Integer getEndLine() {
    return this.endLine;
  }

  public Integer getEndColumn() {
    return this.endColumn;
  }

  public String getFile() {
    return this.file;
  }

  public String getMessage() {
    return this.message;
  }

  public Parser getParser() {
    return this.parser;
  }

  public void setReporter(final String reporter) {
    this.reporter = checkNotNull(reporter, "reporter");
  }

  public String getReporter() {
    return this.reporter;
  }

  public String getRule() {
    return this.rule;
  }

  public String getCategory() {
    return this.category;
  }

  public SEVERITY getSeverity() {
    return this.severity;
  }

  /** The thing that contains the violations. Like a Java/C# class. */
  public String getSource() {
    return this.source;
  }

  /**
   * Some parsers may find values that are specific to that kind of analysis. Those can be made
   * available through this map.
   */
  public Map<String, String> getSpecifics() {
    return this.specifics;
  }

  public Integer getStartLine() {
    return this.startLine;
  }

  public String getGroup() {
    return this.group;
  }

  @Override
  public int hashCode() {
    final int prime = 31;
    int result = 1;
    result = prime * result + (this.category == null ? 0 : this.category.hashCode());
    result = prime * result + (this.column == null ? 0 : this.column.hashCode());
    result = prime * result + (this.endLine == null ? 0 : this.endLine.hashCode());
    result = prime * result + (this.endColumn == null ? 0 : this.endColumn.hashCode());
    result = prime * result + (this.file == null ? 0 : this.file.hashCode());
    result = prime * result + (this.group == null ? 0 : this.group.hashCode());
    result = prime * result + (this.message == null ? 0 : this.message.hashCode());
    result = prime * result + (this.parser == null ? 0 : this.parser.hashCode());
    result = prime * result + (this.reporter == null ? 0 : this.reporter.hashCode());
    result = prime * result + (this.rule == null ? 0 : this.rule.hashCode());
    result = prime * result + (this.severity == null ? 0 : this.severity.hashCode());
    result = prime * result + (this.source == null ? 0 : this.source.hashCode());
    result = prime * result + (this.specifics == null ? 0 : this.specifics.hashCode());
    result = prime * result + (this.startLine == null ? 0 : this.startLine.hashCode());
    return result;
  }

  @Override
  public String toString() {
    return "Violation [column="
        + this.column
        + ", endLine="
        + this.endLine
        + ", endColumn="
        + this.endColumn
        + ", file="
        + this.file
        + ", message="
        + this.message
        + ", parser="
        + this.parser
        + ", reporter="
        + this.reporter
        + ", rule="
        + this.rule
        + ", category="
        + this.category
        + ", severity="
        + this.severity
        + ", source="
        + this.source
        + ", specifics="
        + this.specifics
        + ", startLine="
        + this.startLine
        + ", group="
        + this.group
        + "]";
  }

  @Override
  public int compareTo(final Violation o) {
    return this.comparingString(this).compareTo(this.comparingString(o));
  }

  private String comparingString(final Violation o) {
    final StringBuilder compare = new StringBuilder();
    compare.append(o.file);
    compare.append("_");
    compare.append(Integer.MAX_VALUE - o.getStartLine());
    compare.append("_");
    compare.append(o.getParser());
    compare.append("_");
    compare.append(o.getMessage());
    compare.append("_");
    compare.append(o.getReporter());
    compare.append("_");
    compare.append(o.getEndLine());
    compare.append("_");
    compare.append(o.getEndColumn());
    compare.append("_");
    compare.append(o.getColumn());
    compare.append("_");
    compare.append(o.getSeverity());
    compare.append("_");
    compare.append(o.getSource());
    compare.append("_");
    compare.append(o.getRule());
    compare.append("_");
    compare.append(o.getCategory());
    compare.append("_");
    compare.append(o.getGroup());
    o.getSpecifics().forEach((k, v) -> compare.append("_").append(k).append(v));
    return compare.toString();
  }
}
