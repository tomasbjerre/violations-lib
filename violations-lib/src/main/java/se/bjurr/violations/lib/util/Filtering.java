package se.bjurr.violations.lib.util;

import static se.bjurr.violations.lib.model.SEVERITY.ERROR;
import static se.bjurr.violations.lib.model.SEVERITY.INFO;
import static se.bjurr.violations.lib.model.SEVERITY.WARN;

import java.util.Set;
import java.util.TreeSet;
import se.bjurr.violations.lib.model.SEVERITY;
import se.bjurr.violations.lib.model.Violation;

public final class Filtering {
  private Filtering() {}

  public static Set<Violation> withAtLEastSeverity(
      final Set<Violation> unfiltered, final SEVERITY severity) {
    final Set<Violation> filtered = new TreeSet<>();
    for (final Violation candidate : unfiltered) {
      if (isSeverer(candidate.getSeverity(), severity)) {
        filtered.add(candidate);
      }
    }
    return filtered;
  }

  private static boolean isSeverer(final SEVERITY candiate, final SEVERITY atLeast) {
    return candiate == ERROR //
        || atLeast == INFO //
        || candiate == WARN && atLeast == WARN;
  }
}
