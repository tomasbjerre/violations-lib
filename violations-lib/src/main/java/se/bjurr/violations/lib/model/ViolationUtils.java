package se.bjurr.violations.lib.model;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@SuppressFBWarnings("PATH_TRAVERSAL_IN")
public final class ViolationUtils {

  private ViolationUtils() {}

  public static String relativePath(final List<Path> allFiles, final Violation v) {
    final String file = v.getFile();
    return relativePath(allFiles, file, getUserDir());
  }

  static String relativePath(final List<Path> allFiles, final String file, final String userDir) {
    final String userDirNotNull = Optional.ofNullable(userDir).orElse("");
    final String cwd = Violation.frontSlashes(userDirNotNull);
    final String relativeToCwd = getFileRelativeToCwd(allFiles, file);
    return removeSlashAtBeginning(relativeToCwd, cwd);
  }

  private static String getFileRelativeToCwd(List<Path> allFound, final String file) {
    allFound =
        allFound.stream()
            .filter((it) -> it.toFile().getAbsolutePath().endsWith(file))
            .collect(Collectors.toList());
    if (allFound.isEmpty() || allFound.size() > 1) {
      return file;
    }
    return allFound.get(0).toFile().getAbsolutePath();
  }

  private static String getUserDir() {
    return System.getProperty("user.dir");
  }

  public static List<Path> getAllFiles() {
    return getAllFiles(getUserDir());
  }

  public static List<Path> getAllFiles(final String cwd) {
    if (cwd == null) {
      return new ArrayList<>();
    }
    try (final Stream<Path> stream = Files.walk(Paths.get(cwd))) {
      return stream.filter(Files::isRegularFile).collect(Collectors.toList());
    } catch (final IOException e) {
      return new ArrayList<>();
    }
  }

  private static String removeSlashAtBeginning(final String file, final String cwd) {
    if (file.equals(Violation.NO_FILE)) {
      return file;
    }
    if (cwd == null || cwd.isEmpty()) {
      return removeAnySlashAtBeginning(file);
    }
    final String fileRemoved = removeAnySlashAtBeginning(file);
    final String cwdRemoved = removeAnySlashAtBeginning(cwd);
    if (fileRemoved.startsWith(cwdRemoved)) {
      return removeAnySlashAtBeginning(fileRemoved.substring(cwdRemoved.length()));
    }
    return fileRemoved;
  }

  private static String removeAnySlashAtBeginning(final String file) {
    if (file.startsWith("/")) {
      return file.substring(1);
    }
    return file;
  }
}
