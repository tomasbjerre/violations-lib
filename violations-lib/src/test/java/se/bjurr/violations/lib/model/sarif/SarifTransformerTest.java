package se.bjurr.violations.lib.model.sarif;

import static org.assertj.core.api.Assertions.assertThat;
import static se.bjurr.violations.lib.model.Violation.violationBuilder;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonMappingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.google.gson.GsonBuilder;
import com.networknt.schema.JsonSchema;
import com.networknt.schema.JsonSchemaFactory;
import com.networknt.schema.SpecVersion.VersionFlag;
import com.networknt.schema.ValidationMessage;
import java.io.InputStream;
import java.util.Set;
import java.util.TreeSet;
import org.approvaltests.Approvals;
import org.approvaltests.core.Options;
import org.approvaltests.reporters.AutoApproveReporter;
import org.junit.Test;
import se.bjurr.violations.lib.model.SEVERITY;
import se.bjurr.violations.lib.model.Violation;
import se.bjurr.violations.lib.reports.Parser;

public class SarifTransformerTest {

  @Test
  public void testThatViolationsCanBeTransformed() throws Exception {
    final String description = "asdasd";
    final Integer begin = 123;
    final String path = "/whatever/path.c";
    final Set<Violation> givenViolations = new TreeSet<>();
    final Violation violation1 =
        violationBuilder() //
            .setFile(path) //
            .setMessage(description) //
            .setParser(Parser.CHECKSTYLE) //
            .setRule("Cyclomatic complexity") //
            .setSeverity(SEVERITY.ERROR) //
            .setStartLine(begin) //
            .build();
    givenViolations.add(violation1);
    final Violation violation2 =
        violationBuilder() //
            .setFile(path) //
            .setMessage(description) //
            .setParser(Parser.ANDROIDLINT) //
            .setRule(null) //
            .setSeverity(SEVERITY.INFO) //
            .setStartLine(begin) //
            .build();
    givenViolations.add(violation2);

    final String actual = this.toJson(SarifTransformer.fromViolations(givenViolations));

    this.validateJson(actual);

    Approvals.verify(actual, new Options().withReporter(new AutoApproveReporter()));

    final Set<Violation> parsedViolations =
        Parser.SARIF.getViolationsParser().parseReportOutput(actual, null);

    assertThat(parsedViolations).hasSameSizeAs(givenViolations);
  }

  private void validateJson(final String json)
      throws JsonProcessingException, JsonMappingException {
    final JsonSchemaFactory factory = JsonSchemaFactory.getInstance(VersionFlag.V7);
    final InputStream schemaStream =
        SarifTransformerTest.class.getResourceAsStream("/jsonschemas/sarif-schema.json");
    final JsonSchema jsonSchema = factory.getSchema(schemaStream);
    final JsonNode jsonNode = new ObjectMapper().readTree(json);
    final Set<ValidationMessage> errors = jsonSchema.validate(jsonNode);
    assertThat(errors).as(json).isEmpty();
  }

  private String toJson(final Object o) {
    return new GsonBuilder().setPrettyPrinting().create().toJson(o);
  }
}
