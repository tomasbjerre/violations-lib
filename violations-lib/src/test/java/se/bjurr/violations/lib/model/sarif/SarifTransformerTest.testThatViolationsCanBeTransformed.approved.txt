{
  "version": "2.1.0",
  "runs": [
    {
      "tool": {
        "driver": {
          "name": "Violations Lib",
          "notifications": [],
          "rules": [],
          "taxa": [],
          "locations": [],
          "language": "en-US",
          "contents": [
            "nonLocalizedData"
          ],
          "isComprehensive": true,
          "supportedTaxonomies": []
        },
        "extensions": []
      },
      "invocations": [],
      "language": "en-US",
      "versionControlProvenance": [],
      "artifacts": [],
      "logicalLocations": [],
      "graphs": [],
      "results": [
        {
          "ruleId": "",
          "ruleIndex": -1,
          "kind": "fail",
          "level": "none",
          "message": {
            "text": "asdasd",
            "arguments": []
          },
          "locations": [
            {
              "id": -1,
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "whatever/path.c",
                  "index": -1
                },
                "region": {
                  "startLine": 123,
                  "endLine": 123,
                  "charOffset": -1,
                  "byteOffset": -1,
                  "message": {
                    "text": "asdasd",
                    "arguments": []
                  }
                }
              },
              "logicalLocations": [],
              "annotations": [],
              "relationships": []
            }
          ],
          "stacks": [],
          "codeFlows": [],
          "graphs": [],
          "graphTraversals": [],
          "relatedLocations": [],
          "suppressions": [],
          "rank": -1.0,
          "attachments": [],
          "workItemUris": [],
          "fixes": [],
          "taxa": []
        },
        {
          "ruleId": "Cyclomatic complexity",
          "ruleIndex": -1,
          "kind": "fail",
          "level": "error",
          "message": {
            "text": "asdasd",
            "arguments": []
          },
          "locations": [
            {
              "id": -1,
              "physicalLocation": {
                "artifactLocation": {
                  "uri": "whatever/path.c",
                  "index": -1
                },
                "region": {
                  "startLine": 123,
                  "endLine": 123,
                  "charOffset": -1,
                  "byteOffset": -1,
                  "message": {
                    "text": "asdasd",
                    "arguments": []
                  }
                }
              },
              "logicalLocations": [],
              "annotations": [],
              "relationships": []
            }
          ],
          "stacks": [],
          "codeFlows": [],
          "graphs": [],
          "graphTraversals": [],
          "relatedLocations": [],
          "suppressions": [],
          "rank": -1.0,
          "attachments": [],
          "workItemUris": [],
          "fixes": [],
          "taxa": []
        }
      ],
      "runAggregates": [],
      "redactionTokens": [],
      "newlineSequences": [
        "\r\n",
        "\n"
      ],
      "threadFlowLocations": [],
      "taxonomies": [],
      "addresses": [],
      "translations": [],
      "policies": [],
      "webRequests": [],
      "webResponses": []
    }
  ],
  "inlineExternalProperties": []
}